### PR TITLE
Working Auth System

### PR DESCRIPTION
**Pull request by avitex**
I've made the loading of the user async, to work nicely with the gapi.
If the user has already been signed in, it'll load that profile now rather than having to sign in after a page refresh.

Tell me if you have any other problems :)

### CODE REVIEW

File: src/services/aws.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import AWS from 'aws-sdk/global'
+import 'aws-sdk/clients/dynamodb'
+
+const ROLE_ARN = 'arn:aws:iam::407461997746:role/vueaws'
+
+let AwsDocClient = null
+let awsDocClientResolve = null
+const awsDocClientInit = new Promise((resolve) => awsDocClientResolve = resolve)
```
Comment (avitex) [inline src/services/aws.js]: A little bit hacky, but was an easy solution for making sure the credentials have been set on first page load

Comment (kaihendry) [inline src/services/aws.js]: Yes, that syntax is new to me. I need to read up about "Promises". I am guessing it's like a bash named pipe.

So thinks get triggered only once resolve is called from auth, right?

**Comment by kaihendry** [comment]
Awesome, awesome stuff. Hopefully this code sample will help others.

Not to throw a spanner in the works, but for developing/debugging, I find it simple useful to have the AWS object on window so I can play/experiment with it via devtools. Hopefully this won't sound like undoing your work, but would I just attach it on window from one of my views?

Another issue, is how to log out? :grimacing:

**Comment by avitex** [comment]
I'll make another commit, need to refactor out that dirty promise hack

File: index.html [hunk 2]
```diff
@@ -4,7 +4,7 @@
 		<meta charset="utf-8">
 		<title>AWS VUE</title>
 		<meta name="viewport" content="initial-scale=1, maximum-scale=1, user-scalable=no, minimal-ui">
-		<script src="https://apis.google.com/js/platform.js"></script>
+		<script src="https://apis.google.com/js/platform.js?onload=bootloader" async defer></script>
```
Comment (avitex) [inline index.html]: Didn't mean to leave that in there

File: src/components/SignIn.vue [hunk 3]
```diff
@@ -1,24 +1,23 @@
...
 
 <script>
+import { signIn } from '../services/auth'
+
 export default {
 	data: () => ({
 		error: null
 	}),
-	mounted() {
-		gapi.signin2.render(this.$refs.test, {
```
Comment (avitex) [inline src/components/SignIn.vue]: This method doesn't seem to be very friendly with Vue, keeps throwing a style error on router change, so you'll have to style the signin button yourself.

File: src/services/aws.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+export function setCredentials(idToken) {
+	AWS.config.credentials = new AWS.WebIdentityCredentials({
+		RoleArn: ROLE_ARN,
+		WebIdentityToken: idToken,
+	})
+
+	AwsDocClient = new AWS.DynamoDB.DocumentClient({
+		region: 'ap-southeast-1',
+	})
```
Comment (avitex) [inline src/services/aws.js]: @kaihendry For debugging, you could just set window.awsClient here

**Comment by kaihendry** [comment]
Hmmm, hitting this issue. http://s.natalian.org/2016-10-21/1477028663_2174x1404.png

Must be some cookie or something messing about since it works fine in Incognito!

File: index.html (hunk 2 above)
Comment (kaihendry) [inline index.html]: the `onload=bootloader` part?

**Comment by avitex** [comment]
I just realized I left  `async, defer` in the script include, the GAPI isn't probably available when it attempts to run

File: index.html (hunk 2 above)
Comment (avitex) [inline index.html]: Yeah

**Comment by avitex** [comment]
A small note to add would be, if you are planning to run this on IE or older browsers, you'll want to add a promise polyfill

**Pull request merged**